cli/spinner.ts

```diff
@@ -0,0 +1,163 @@
+}
+
+/**
+ * A spinner that can be used to indicate that something is loading.
+ */
+export class Spinner {
+  #message: string | undefined;
+  #speed: number;
+  #spinner: string[];
+  #color: string | undefined;
+  #cleanup: boolean;
+  
+  #intervalId: number | undefined;
+
+  /**
+   * Creates a new spinner.
+   * @param message - The message to display next to the spinner.
+   * @param speed - The speed of the spinner.
+   * @param spinner - The spinner to use.
+   * @param color - The color of the spinner.
+   * @param cleanup - Whether to cleanup the output if the process is terminated.
+   */
+  constructor(message: string | undefined, speed: number, spinner: string[], color: string | undefined, cleanup: boolean) {
+    this.#message = message;
+    this.#speed = speed;
+    this.#spinner = spinner;
+    this.#color = color;
+    this.#cleanup = cleanup;
+  }
+
+  /**
+   * Starts the spinner.
+   */
+  start() {
+    let i = 0;
+    const message = this.#message ?? '';
+    
+    const color = this.#color ?? '';
+
+    if (Deno.stdout.writable.locked) return;
+
+    // Updates the spinner after the given interval.
+    const updateFrame = () => {
+      Deno.stdout.writeSync(LINE_CLEAR);
+      let frame;
+      // If color was specified, add it to the frame. Otherwise, just use the spinner and message.
+      if (this.#color) frame = encoder.encode(color + this.#spinner[i] + COLOR_RESET + ' ' + message);
+      else frame = encoder.encode(this.#spinner[i] + ' ' + message);
+      Deno.stdout.writeSync(frame);
+      i = (i + 1) % this.#spinner.length;
+    }
+
+    this.#intervalId = setInterval(updateFrame, this.#speed);
+
+    // Cleanup if the process is terminated.
+    if (this.#cleanup) Deno.addSignalListener("SIGINT" , this.cleanup.bind(this));
+  }
+
+  /**
+   * Stops the spinner.
+   */
+  stop() {
+    if (this.#intervalId) {
+      clearInterval(this.#intervalId);
+      Deno.stdout.writeSync(LINE_CLEAR); // Clear the current line
+    }
+  }
+  
+  /**
+   * Stops the spinner and cleans up the output.
+   */
+  cleanup() {
+    this.stop();
+    Deno.stdout.writeSync(encoder.encode("\n"));
+  }
+}
+
+/**
+ * Creates a spinner with the given options.
+ * 
+ * Note: The spinner will _not_ start automatically.
+ * @param options - The options for the spinner. 
+ * @returns The created spinner.
+ * 
+ * @example 
+ * ```ts
+ * const mySpinner = createSpinner({
+ *   spinner: spinners.lineSpinner,
+ *   message: 'Loading...',
+ * });
+ * 
+ * mySpinner.start();
+ * ```
+ * 
+ * You can also add custom spinners:
+ * ```ts
+ * const mySpinner = createSpinner({
+ *  spinner: ['1', '2', '3', '4', '5', '6', '7', '8', '9'],
+ *  message: 'Loading...',
+ *  speed: 20
+ * });
+ * 
+ * mySpinner.start();
+ * ```
+ */
+export function createSpinner({ message, speed = DEFAULT_SPEED, spinner, color, cleanup = true }: SpinnerOptions): Spinner {
+  return new Spinner(message, speed, spinner, color, cleanup);
+}
+
```

Comment:
I don't think this function is needed as it's just a wrapper around the `Spinner` constructor.